runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/MSecMonitoringInfoToCounterUpdateTransformer.java

```diff
@@ -0,0 +1,139 @@
+import java.util.Map;
+import java.util.Optional;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.MonitoringInfo;
+import org.apache.beam.runners.core.metrics.SpecMonitoringInfoValidator;
+import org.apache.beam.runners.dataflow.worker.DataflowExecutionContext.DataflowStepContext;
+import org.apache.beam.runners.dataflow.worker.counters.DataflowCounterUpdateExtractor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MSecMonitoringInfoToCounterUpdateTransformer
+    implements MonitoringInfoToCounterUpdateTransformer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BeamFnMapTaskExecutor.class);
+
+  private SpecMonitoringInfoValidator specValidator;
+  private Map<String, DataflowStepContext> transformIdMapping;
+  private Map<String, String> urnToCounterNameMapping;
+
+  /**
+   * @param specValidator SpecMonitoringInfoValidator to utilize for default validation.
+   * @param transformIdMapping Mapping of PTransform ID string to DataflowStepContext.
+   */
+  public MSecMonitoringInfoToCounterUpdateTransformer(
+      SpecMonitoringInfoValidator specValidator,
+      Map<String, DataflowStepContext> transformIdMapping) {
+    this.specValidator = specValidator;
+    this.transformIdMapping = transformIdMapping;
+    urnToCounterNameMapping = createKnownUrnToCounterNameMapping();
+  }
+
+  /** Allows to inject members for cleaner testing. */
+  @VisibleForTesting
+  protected MSecMonitoringInfoToCounterUpdateTransformer(
+      SpecMonitoringInfoValidator specValidator,
+      Map<String, DataflowStepContext> transformIdMapping,
+      Map<String, String> urnToCounterNameMapping) {
+    this.specValidator = specValidator;
+    this.transformIdMapping = transformIdMapping;
+    this.urnToCounterNameMapping = urnToCounterNameMapping;
+  }
+
+  @VisibleForTesting
+  protected Map<String, String> createKnownUrnToCounterNameMapping() {
+    Map<String, String> result = new HashMap<>();
+    result.put("beam:metric:pardo_execution_time:start_bundle_msecs:v1", "start-msecs");
+    result.put("beam:metric:pardo_execution_time:process_bundle_msecs:v1", "process-msecs");
+    result.put("beam:metric:pardo_execution_time:finish_bundle_msecs:v1", "finish-msecs");
+    return result;
+  }
+
+  /**
+   * Validates provided monitoring info against specs and common safety checks.
+   *
+   * @param monitoringInfo to validate.
+   * @return Optional.empty() all validation checks are passed. Optional with error text otherwise.
+   * @throws RuntimeException if received unexpected urn.
+   */
+  protected Optional<String> validate(MonitoringInfo monitoringInfo) {
+    Optional<String> validatorResult = specValidator.validate(monitoringInfo);
+    if (validatorResult.isPresent()) {
+      return validatorResult;
+    }
+
+    String urn = monitoringInfo.getUrn();
+    if (!urnToCounterNameMapping.keySet().contains(urn)) {
+      throw new RuntimeException(String.format("Received unexpected counter urn: %s", urn));
```

Comment:
Please just drop metrics if they are invalid, don't throw an exception. This would be bad for many of our users, who don't care about metrics but just want their pipeline to finish properly.

Reply:
This exception doesn't validate MonitoringInfo, instead it checks validity of logic. 
If MSecMonitoringInfoToCounterUpdateTransformer receives non-supported counter it is not a problem of counter, but invoking code.